text/0000-fabric-private-chaincode.md

```diff
@@ -0,0 +1,472 @@
+
+Note on Terminology: The current feature naming scheme includes several elements that originated in Intel SGX and therefore use the formerly proprietary term Enclave rather than TEE. Earlier versions of this RFC stated an aim to replace the term Enclave with TEE, but since then the two terms have come to be accepted as interchangeable. We therefore decided not to try to expunge the term, but to use it as the inductry has begun to do, to refer to various TEEs. The project team currently participates in the new Confidential Computing Consortium which aims to promote standards for deployment of workloads across TEEs/Enclaves, and we intend to align with their terminology as it evolves.
+
+
+## Principal Components of FPC:
+
+![Architecture](../images/fpc/high-level/FPC-Architecture.png)
+
+**FPC Chaincode:** A chaincode created by a developer to run in a Chaincode Enclave.
+Unlike regular Fabric chaincode, an FPC Chaincode must currently be written in C++ using our FPC SDK. A future goal for the project is to support additional languages, e.g., by the use of WebAssembly.
+
+**Chaincode Enclave:** An enclave in which a single particular chaincode executes.
+The Chaincode Enclave contains the chaincode (called an FPC Chaincode) to be executed, with its linked FPC Shim. The FPC runtime is responsible for creating the Chaincode Enclave when the FPC chaincode is installed on the endorsing peers.
+
+**FPC Shim:** This is the externally exposed shim interface to an FPC Chaincode.
+The FPC Shim is composed of two parts called "stubs", one residing inside the enclave (Referred to in our diagrams as "FPC_stub_enclave") and another one outside the enclave (Referred to as "FPC_stub"). The portion of the shim interface exposed to the FPC chaincode inside the enclave is written primarily in C++, whereas the external counterpart is written in Go and mainly responsible for communication between the Peer and the FPC chaincode.
+The FPC Shim enables an FPC Chaincode to talk to the Peer, in particular to invoke the chaincode and provide "normal" shim operations such as `getState` and `putState`.
+The FPC SDK provides the FPC Shim to enable FPC chaincode developers to easily link to this library when compiling their FPC Chaincode and building a deployment artifact. It implements the interface from the Peer to the FPC Chaincode, encrypts and decrypts state and arguments, and implements Attestation.
+See the [FPC Shim](#fpc-shim) section for more information.
+
+**Ledger Enclave:** The Ledger Enclave is a crucial component to protect the execution of a FPC chaincode.
```

Comment:
Why do you need the ledger enclave? Can't you sign each (key, value, version) in the read-write set and then have the FPC always verify the signature whenever it reads the value during computation?

Reply:
you need consistency of the whole view and history. Strictly speaking you don't need the trusted ledger and could validate the history back to the genesis block but obviously that would be extremely inefficient. In essence, the trusted ledger is a cache of evaluted/validated history (as it is for standard fabric, except in this case with a different thread/trust model)